### DISCUSSION TITLE
.d.ts Generation?

### DISCUSSION THREAD

**Discussion by GEEKiDoS** [sebastienros/jint#1244] [Ideas]
It would be nice if there's a typescript type declaration generator for CLR types

**Comment by lahma** [comment]
Would you like to try to implement one? Generally this kind of functionality is out-of-scope for Jint.

**Comment by Konijima** [comment]
@GEEKiDoS I use https://github.com/DogusTeknoloji/cs-to-ts with some modification it works pretty well.

Ouput example:
```
declare namespace LemonUI.Extensions {
    abstract class SizeExtensions {
        static ToRelative(size: SizeF): SizeF;
        static ToAbsolute(size: SizeF): SizeF;
    }
}
```
Note I have modified the CsToTs engine to suit my project.

**Reply by XuJin186** [reply]
Can your modified version be viewed on github?

**Comment by hyzx86** [comment]
Sharing a repository, while Jint doesn't support Typescript directly, we can write Typescript to translate it to Javascript

https://fabiandev.github.io/typescript-playground/?ts=nightly

**Reply by cyraid** [reply]
> Sharing a repository, while Jint doesn't support Typescript directly, we can write Typescript to translate it to Javascript
> 
> https://fabiandev.github.io/typescript-playground/?ts=nightly

@hyzx86 I believe they're referring to accessing the C# classes in say, an editor using Typescript and .d.ts files so they could have code completion.

**Reply by Xicy** [reply]
I have a workaround for running typescript in Jint. It's not an efficient way but it works.

Below code, block add the ability to the transformation of ts to js.
```cs
engine.Realm.GlobalObject.FastAddProperty("window", engine.Realm.GlobalObject, false, false, false);
engine.Realm.GlobalObject.FastAddProperty("global", engine.Realm.GlobalObject, false, false, false);
engine.Realm.GlobalObject.FastAddProperty("transpileModule", engine.ImportModule("tsc.esm.min.js").Get("transform"), false, false, false);
engine.Realm.GlobalObject.FastAddProperty("transformPre", engine.Evaluate(@"(file,fileName)=>transpileModule(file, (ts)=> ({
        fileName,
        compilerOptions: {
          target: ts.ScriptTarget.ES5,
          module: ts.ModuleKind.ES2015,
          jsx: ts.JsxEmit.React,
          isolatedModules: true,
          esModuleInterop: true,
          types: [],
          lib: ['ES2015.Promise'],
          newLine: ts.NewLineKind.LineFeed,
          allowJs: true,
          sourceMap: false,
          downlevelIteration: true,
          importHelpers: true,
          inlineSources:false,
          removeComments:true,
          experimentalDecorators:true,
          emitDecoratorMetadata:true
    },
    reportDiagnostics: false
}))"), false, false, false);
```

Then you need to implement IModuleLoader for loading ts modules 
```cs
public Module LoadModule(Engine engine, ResolvedSpecifier resolved)
    {
        var file = fileSystem.GetFileInfo(resolved.Uri.AbsolutePath);
        var reader = new StreamReader(file.CreateReadStream(), Encoding.UTF8);
        var result = reader.ReadToEnd();
        
        if (!resolved.Uri.AbsolutePath.Contains("tsc.esm.min.js"))
        {
            var a = engine.Call("transformPre", result, resolved.Uri.LocalPath);
            result = a.AsString();
        }
        
        var parserOptions = new JsxParserOptions()
        {
            AdaptRegexp = true,
            Tolerant = true,
        };
        return new JsxParser(parserOptions).ParseModule(result, resolved.Uri.LocalPath)!;
    }
```

tsc.esm.min.js is Typescript 5.0 preview version that has custom bundling for removing some node dependencies like the filesystem, performance counter, etc.
If you want to generate your self you can take a look [this repository](https://github.com/viceice/dotnet-ts-testing/blob/main/rollup.config.mjs) 
[tsc.esm.min.js.txt](https://github.com/sebastienros/jint/files/11008271/tsc.esm.min.js.txt)

**Reply by Xicy** [reply]
I think, I misunderstood the idea, for type generations Else Workflow used some reflection things I did not have a time to analyze it.

Else can create type declaration for UI side. Its use monaco-editor and provide a d.ts file.
![image](https://user-images.githubusercontent.com/8670548/226101427-c2aceeca-b949-4529-a5b3-f68239171e67.png)


https://github.com/elsa-workflows/elsa-core/blob/master/src/scripting/Elsa.Scripting.JavaScript/Handlers/RenderJavaScriptTypeDefinitions.cs

**Reply by hyzx86** [reply]
@Xicy that feature is from `monaco-editor` 

just like this :
```ts

    defaultResult.get('/api/ScriptMethod/GetTsSuggestion').then(response => {
      let suggestion = response.data?.data;

      monaco.languages.typescript.typescriptDefaults.addExtraLib(suggestion, 'getTsSuggestion.d.ts');
      monaco.languages.typescript.javascriptDefaults.addExtraLib(suggestion, 'getTsSuggestion.d.ts');
    });
```
